clang/test/CodeGenSYCL/sycl-device-alias.cpp

```diff
@@ -0,0 +1,45 @@
+// RUN: %clang_cc1 -fsycl -fsycl-is-device -triple spir64-unknown-unknown-sycldevice -disable-llvm-passes -emit-llvm %s -o - | FileCheck %s
+// Test that aliasing does not force an unused entity to be emitted
+
+// CHECK-NOT: define spir_func void @unused_func()
+extern "C" void unused_func() {}
+// CHECK-NOT: @unused_aliaser
+extern "C" void unused_aliaser() __attribute__((alias("unused_func")));
+// CHECK-NOT: @unused_int
+int unused_int = 3;
+// CHECK-NOT: @alias_unused_int
+extern int alias_unused_int __attribute__((alias("unused_int")));
+
+// CHECK-DAG: define spir_func void @used_func()
+extern "C" void used_func() {}
+// CHECK-DAG: @aliaser = alias void (), void ()* @used_func
+extern "C" void aliaser() __attribute__((alias("used_func")));
+
+// CHECK-DAG: define spir_func void @func()
+extern "C" void func() {}
+// CHECK-DAG: @used_aliaser = alias void (), void ()* @func
+extern "C" void used_aliaser() __attribute__((alias("func")));
+
+// CHECK-DAG: @used_int = addrspace(1) constant i32 5, align 4
+extern "C" const int used_int = 5;
+// CHECK-DAG: @alias_used_int = alias i32, addrspacecast (i32 addrspace(1)* @used_int to i32*)
```

Comment:
Wow, sorry, but I've just notified that we have incorrect address space cast here. It casts from global (1) address space to private (no number or 0), this is not correct. We can cast only non-constant address space (1,3,0 in spir targets) to generic address space (4 in spir target).

Reply:
Thanks for catching this; this isn't due to my change, but I can change that if it is broken.   What needs to happen here?

Also, is 0 private or just the default?   I thought private was 4.

Reply:
It looks like 'used_int' is allocated in addrspace 1; and when the alias is created, it doesn't consider the addrspace of the aliasee (but rather the address space of its type).   I will look into changing that.

Reply:
> Also, is 0 private or just the default? I thought private was 4.

There are difference between AST address spaces and IR address spaces. So, OpenCL/SYCL has global, constant, local, private and generic address spaces. They are mapped to some numbers in IR. These numbers actually depend on target. We mostly use SPIR-based target. So, in SPIR targets 1 is global, 2 is constant, 3 is local, 4 is generic, 0 is private. In AST we also have address space qualifiers, and in our SYCL implementation if pointer has some AS qualifier, it will have this same qualifier in AST. I.e. if it was expliticly declared, i.e. `__global int *ptr` etc, this pointer will have global AS qualifier in AST and it will be codegen-ed with address space "1". But, if some pointer doesn't have address space qualifier it will have some **default** address space in AST. Here is the thing that two pointers with different address spaces but with the same type are considered as pointers of different type (i.e. `__global int*` and `__local int*` are different types), but, In SYCL user doesn't have to define different overload for each function but with different address spaces in parameters, so the conversion should be done by the compiler. So, our solution was to map **default** AST address space to **generic** address space in SPIR target, because it is valid to cast from most of address spaces to generic. So, for each place where **default** address space was in AST, we emit **generic** (4 in SPIR target), so the following tools and backends read it as **generic**. If I remember correct the only thing that we left in SPIR private address space is device-side allocas, because it is not possible to allocate with generic address space.

Reply:
@Fznamznon, please take a look.  I think I have addressed this issue.